src/ja/common/work_machine.py

```diff
@@ -15,28 +15,86 @@ def __init__(self, cpu_threads: int, memory: int, swap: int):
         @param memory The amount of RAM, in MB.
         @param swap The amount of swap space, in MB.
         """
+        if cpu_threads < 0:
+            raise ValueError("CPU threads cannot be negative")
+        if memory < 0:
+            raise ValueError("Memory cannot be negative")
+        if swap < 0:
+            raise ValueError("Swap cannot be negative")
+
+        self._cpu_threads = cpu_threads
+        self._memory = memory
+        self._swap = swap
 
     @property
     def cpu_threads(self) -> int:
         """!
         @return The allocated CPU threads.
         """
+        return self._cpu_threads
 
     @property
     def memory(self) -> int:
         """!
         @return The allocated RAM in MB.
         """
+        return self._memory
 
     @property
     def swap(self) -> int:
         """!
         @return The allocated swap space in MB.
         """
+        return self._swap
+
+    def is_negative(self) -> bool:
+        """
+        @returns bool indicating if any of the attributes is negative
+        """
+        ret: bool = self._cpu_threads < 0 or self._memory < 0 or self._swap < 0
+        return ret
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, ResourceAllocation):
+            return self._cpu_threads == other.cpu_threads and self._memory == other.memory and self._swap == other.swap
+        else:
+            return False
+
+    def __sub__(self, other: ResourceAllocation) -> "ResourceAllocation":
+        cpu_threads: int = self._cpu_threads - other.cpu_threads
+        memory: int = self._memory - other.memory
+        swap: int = self._swap - other.swap
+
+        return ResourceAllocation(cpu_threads, memory, swap)
+
+    def __add__(self, other: ResourceAllocation) -> "ResourceAllocation":
+        cpu_threads: int = self._cpu_threads + other.cpu_threads
+        memory: int = self._memory + other.memory
+        swap: int = self._swap + other.swap
+
+        return ResourceAllocation(cpu_threads, memory, swap)
+
+    def __isub__(self, other: ResourceAllocation) -> "ResourceAllocation":
+        self._cpu_threads -= other.cpu_threads
+        self._memory -= other.memory
+        self._swap -= other.swap
+
+        return self
+
+    def __iadd__(self, other: ResourceAllocation) -> "ResourceAllocation":
+        self._cpu_threads += other.cpu_threads
+        self._memory += other.memory
+        self._swap += other.swap
+
+        return self
 
     def to_dict(self) -> Dict[str, object]:
-        pass
+        return self.__dict__
```

Comment:
Please replace this with a manual implementation where you use the properties instead of the object's attributes.

Reply:
I think it might be ok for such small classes? I mean, we know what is going to go out of the class. We don't have nesting either.

Reply:
One of the consequences of using __dict__ is that underscores get added to property names.
This means that a user writing a config file would have to guess whether to put an underscore in front of a property name unless we were to put underscores in front of all property names (I don't think we should).